File: sdk-python/packages/sync/src/airprompter_agent_sync/sync/pull_bundle.py
"""One pull, no store: fetch the release promoted to a target, run the whole verification chain, and hand back an
``.apbundle`` — the artifact a fleet carries through its own store (a database row, an object, a config entry) to
runtimes that never hold an Agent key. ``airprompter pull`` is this function with files around it; a puller job is
this function with a database around it. The bundle is sealed to the fleet's distribution key unless the caller asks
for plaintext, which only the dev target permits.

The cheap path: the control plane names an edge pointer (a few hundred bytes behind a CDN, ``generation.json``) in
its manifest answer; a puller that hands back ``edge`` from the last result reads the pointer first — a 304, or a
generation it already holds, means nothing moved and the origin is never called. Only a moved pointer (or none known)
reaches the API, and that read is conditional too. Steady state costs a CDN 304 per interval, not an API request;
``next_pull_delay_ms`` stretches the interval while nothing changes."""

from __future__ import annotations

import base64
from dataclasses import dataclass, field
from datetime import datetime, timedelta, timezone
from typing import Any, Callable, Mapping, Optional

from airprompter_agent_core.bundle.apbundle import create_encrypted_bundle, create_plaintext_bundle
from airprompter_agent_core.control.client import SyncClient
from airprompter_agent_core.protocol.trust import referenced_payloads, verify_manifest, verify_root_metadata


@dataclass
class PullEdgeState:
    """The puller's memory between pulls: content-free, safe to persist beside the table."""

    pointer_url: Optional[str] = None
    pointer_etag: Optional[str] = None
    manifest_etag: Optional[str] = None


def next_pull_delay_ms(*, outcome: str, unchanged_streak: int, interval_ms: int, cap_ms: int = 5 * 60 * 1000) -> int:
    """How long to wait before the next pull: the interval while things change, doubling while nothing does, never past
    ``cap_ms``; back to the interval on any change, refusal or outage — those are what a puller is for."""
    if outcome != "unchanged":
        return interval_ms
    stretched = interval_ms * 2 ** min(unchanged_streak, 20)
    return min(max(interval_ms, stretched), max(cap_ms, interval_ms))


@dataclass
class PullBundleResult:
    status: str  # "ok" | "unchanged" | "nothing_promoted" | "refused" | "unavailable"
    #: For "unchanged": "pointer" (a CDN read, no origin call) or "origin" (the API's 304).
    via: Optional[str] = None
    #: Hand this back as ``edge`` on the next pull.
    edge: PullEdgeState = field(default_factory=PullEdgeState)
    bundle: Optional[dict[str, Any]] = None
    manifest: Optional[dict[str, Any]] = None
    generation: Optional[int] = None
    release_digest: Optional[str] = None
    created_at: Optional[str] = None
    not_after: Optional[str] = None
    trusted_root: Optional[Mapping[str, Any]] = None
    reason: Optional[str] = None
    content_hash: Optional[str] = None
    detail: Optional[str] = None
    held: Optional[int] = None
    extra: dict[str, Any] = field(default_factory=dict)


def _iso(dt: datetime) -> str:
    return dt.astimezone(timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z")


def pull_bundle(
    *,
    client: SyncClient,
    scope: Mapping[str, str],
    trusted_root: Mapping[str, Any],
    now: Callable[[], str],
    distribution_public_key: Optional[bytes],
    fetch_root: Optional[Callable[[], Optional[Mapping[str, Any]]]] = None,
    minimum_generation: int = 0,
    not_after_days: float = 90,
    countersign_root: Optional[Mapping[str, Any]] = None,
    require_countersign: Optional[bool] = None,
    edge: Optional[PullEdgeState] = None,
    skip_pointer: bool = False,
) -> PullBundleResult:
    """``distribution_public_key``: the fleet's X25519 public key (32 raw bytes) the bundle is sealed to; ``None`` writes
    plaintext, allowed for the dev target only. ``fetch_root``: the environment's root document — a pinned key alone
    names the ROOT, not the signing keys under it, so without it every manifest refuses ``unknown_signing_key`` unless
    ``trusted_root`` is already a full document. ``minimum_generation``: the newest generation the caller already holds;
    an older answer is reported as ``generation_rollback`` instead of becoming a quiet extra row. ``edge``: what the last
    result handed back (the pointer URL the control plane named and the ETags); ``skip_pointer``: a nudge said "check
    now" — read the origin, conditionally."""
    at = now()
    state = PullEdgeState(edge.pointer_url, edge.pointer_etag, edge.manifest_etag) if edge else PullEdgeState()
    if distribution_public_key is None and scope["target"] != "dev":
        return PullBundleResult(status="refused", reason="plaintext_not_allowed", edge=state)
    if not (0 < not_after_days <= 365):
        raise ValueError("not_after_days must be a positive number of days, 365 at most")
    root = trusted_root
    try:
        # The pointer first: unsigned and cacheable, it can only say "nothing moved" — never extend trust. A 304, or a
        # generation the caller already holds, ends the pull at the CDN. Anything else goes on to the origin.
        if state.pointer_url and not skip_pointer:
            pointer = client.edge_pointer(state.pointer_url, state.pointer_etag)
            if pointer.status == "not_modified":
                return PullBundleResult(status="unchanged", via="pointer", edge=state)
            if pointer.status == "ok":
                state.pointer_etag = pointer.etag
                if minimum_generation > 0 and int((pointer.pointer or {}).get("generation", 0)) <= minimum_generation:
                    return PullBundleResult(status="unchanged", via="pointer", edge=state)
        if fetch_root is not None:
            candidate = fetch_root()
            if candidate is not None:
                verdict = verify_root_metadata(candidate=candidate, trusted=root, now=at)
                # A root that does not descend from the trusted one is the finding, not a detail behind unknown_signing_key.
                if not verdict.ok:
                    return PullBundleResult(status="refused", reason="root_refused", detail=verdict.reason, edge=state)
                root = candidate
        # Conditional: the origin answers 304 to the ETag it last gave, and names the pointer either way.
        fetched = client.manifest(if_none_match=state.manifest_etag)
        if fetched.status == "not_found":
            return PullBundleResult(status="nothing_promoted", edge=state)
        if fetched.status == "unauthorized":
            return PullBundleResult(status="unavailable", reason="unauthorized", edge=state)
        if fetched.status == "forbidden":
            return PullBundleResult(status="unavailable", reason="forbidden", detail=fetched.code, edge=state)
        if fetched.status == "error":
            return PullBundleResult(status="unavailable", reason=f"http_{fetched.http_status}", edge=state)
        if fetched.edge_pointer_url:
            state.pointer_url = fetched.edge_pointer_url
        if fetched.status == "not_modified":
            return PullBundleResult(status="unchanged", via="origin", edge=state)
        state.manifest_etag = fetched.etag
        manifest = fetched.manifest
        assert manifest is not None
        if int(manifest["payload"]["generation"]) < minimum_generation:
            return PullBundleResult(status="refused", reason="generation_rollback", detail=f"the control plane answered generation {manifest['payload']['generation']}; the caller holds {minimum_generation}", held=minimum_generation, edge=state)

        payloads: dict[str, bytes] = {}
        for content_hash, byte_length in referenced_payloads(manifest["payload"]).items():
            data = client.payload(content_hash)
            if data is None:
                return PullBundleResult(status="refused", reason="payload_missing", content_hash=content_hash, edge=state)
            if len(data) != byte_length:
                return PullBundleResult(status="refused", reason="payload_length_mismatch", content_hash=content_hash, edge=state)
            payloads[content_hash] = data
        # No stored generation here: a puller has no host to move backwards. Anti-rollback is the store's rule, applied
        # by every runtime that opens this bundle.
        verdict = verify_manifest(manifest=manifest, root=root, now=at, scope=scope, stored_generation=0, payloads=payloads, countersign_root=countersign_root, require_countersign=require_countersign)
        if not verdict.ok:
            return PullBundleResult(status="refused", reason=verdict.reason, edge=state)

        not_after = _iso(datetime.fromisoformat(at.replace("Z", "+00:00")) + timedelta(days=not_after_days))
        contents = {
            "createdAt": at,
            "notAfter": not_after,
            "manifest": manifest,
            "keySet": root,
            "payloads": [{"contentHash": h, "byteLength": len(b), "bytes": base64.urlsafe_b64encode(b).rstrip(b"=").decode("ascii")} for h, b in payloads.items()],
        }
        bundle = create_encrypted_bundle(contents, distribution_public_key) if distribution_public_key is not None else create_plaintext_bundle(contents)
        return PullBundleResult(
            status="ok",
            bundle=bundle,
            manifest=manifest,
            generation=int(manifest["payload"]["generation"]),
            release_digest=str(manifest["payload"]["releaseDigest"]),
            created_at=at,
            not_after=not_after,
            trusted_root=root,
            edge=state,
        )
    except Exception as error:  # noqa: BLE001 — a puller reports, it does not crash the job
        return PullBundleResult(status="unavailable", reason="network", detail=str(error), edge=state)

File: sdk-typescript/packages/sync/src/sync/pullBundle.ts
/**
 * One pull, no store: fetch the release promoted to a target, run the whole
 * verification chain, and hand back an `.apbundle` — the artifact a fleet
 * carries through its own store (a database row, an object, a config
 * entry) to runtimes that never hold an Agent key. `airprompter pull` is
 * this function with files around it; a puller job is this function with a
 * database around it. The bundle is sealed to the fleet's distribution key
 * unless the caller asks for plaintext, which only the dev target permits.
 *
 * The cheap path: the control plane names an edge pointer (a few hundred
 * bytes behind a CDN, `generation.json`) in its manifest answer; a puller
 * that hands back `edge` from the last result reads the pointer first — a
 * 304, or a generation it already holds, means nothing moved and the origin
 * is never called. Only a moved pointer (or none known) reaches the API,
 * and that read is conditional too (`manifestEtag`). Steady state costs a
 * CDN 304 per interval, not an API request; `nextPullDelayMs` stretches the
 * interval while nothing changes.
 */

import { createEncryptedBundle, createPlaintextBundle, referencedPayloads, verifyManifest, verifyRootMetadata } from "@airprompter/agent-core";
import type { Bundle, BundleContents, Manifest, RefusalCode, RootMetadata, SyncClient } from "@airprompter/agent-core";

export interface PullBundleInput {
  client: SyncClient;
  scope: { organizationId: string; agentId: string; target: "dev" | "staging" | "prod" };
  /** The last accepted root, or the synthetic pinned document. */
  trustedRoot: RootMetadata;
  /**
   * The environment's root document, fetched beside the manifest. A pinned key alone names the ROOT, not the signing
   * keys under it: with `trustedRoot` built from a pinned key and no `fetchRoot`, every manifest refuses
   * `unknown_signing_key`. Pass the fetch (the edge serves `roots/{environment}/root.json`) unless `trustedRoot` is
   * already a full root document.
   */
  fetchRoot?: () => Promise<RootMetadata | null>;
  /**
   * The newest generation the caller already holds (its store's MAX(generation)). A control plane answering with an
   * OLDER generation is reported as `generation_rollback` instead of becoming a quiet extra row. 0 when unknown.
   */
  minimumGeneration?: number;
  now: () => string;
  /**
   * The fleet's X25519 distribution public key (32 raw bytes): the bundle is sealed to it. `null` writes plaintext,
   * allowed for the dev target only — every other target's bundle is ciphertext wherever it lands.
   */
  distributionPublicKey: Uint8Array | null;
  /** How long the bundle stays usable as a fallback; 90 days unless told otherwise, 365 at most. */
  notAfterDays?: number;
  countersignRoot?: RootMetadata | null;
  requireCountersign?: boolean;
  /**
   * What the last result handed back: the pointer URL the control plane named, the pointer's ETag, and the ETag of the
   * manifest last read. Pass it back verbatim; the first pull has none and goes to the origin once.
   */
  edge?: PullEdgeState | null;
  /** Skip the pointer for this pull (a nudge said "check now", or the operator asked): the origin is read, conditionally. */
  skipPointer?: boolean;
  /**
   * The bound on a stuck pointer: when the origin was last consulted longer ago than this, the pointer is skipped and
   * the origin read (conditionally — one API 304 when nothing moved). A pointer a CDN pinned, or one the control
   * plane failed to write, can hide a promotion only this long. One hour unless told otherwise.
   */
  maxPointerAgeMs?: number;
}

/**
 * The puller's memory between pulls: content-free. Persist it WITH the row it was returned beside (the same
 * transaction), never before it — a saved `manifestEtag` for a row that was never written makes the next origin read
 * a 304 and the row is never written.
 */
export interface PullEdgeState {
  pointerUrl: string | null;
  pointerEtag: string | null;
  manifestEtag: string | null;
  /** When the origin last answered (ISO); the pointer is trusted to say "nothing moved" only for `maxPointerAgeMs` after it. */
  lastOriginAt: string | null;
}

const NO_EDGE: PullEdgeState = { pointerUrl: null, pointerEtag: null, manifestEtag: null, lastOriginAt: null };
export const DEFAULT_MAX_POINTER_AGE_MS = 60 * 60 * 1000;

export type PullBundleResult =
  | { status: "ok"; bundle: Bundle; manifest: Manifest; generation: number; releaseDigest: string; createdAt: string; notAfter: string; trustedRoot: RootMetadata; edge: PullEdgeState }
  /** Nothing moved: the pointer (a CDN read, no origin call) or the origin's 304 said so. */
  | { status: "unchanged"; via: "pointer" | "origin"; edge: PullEdgeState }
  | { status: "nothing_promoted"; edge: PullEdgeState }
  | { status: "refused"; reason: RefusalCode | "payload_missing" | "payload_length_mismatch" | "plaintext_not_allowed" | "root_refused" | "generation_rollback"; contentHash?: string; detail?: string; held?: number; edge: PullEdgeState }
  | { status: "unavailable"; reason: "unauthorized" | "forbidden" | "network" | `http_${number}`; detail?: string; edge: PullEdgeState };

/**
 * How long to wait before the next pull: the interval while things change, stretching by doubling while nothing does
 * (a pointer 304 costs little, but a thousand pullers asking every ten seconds for a release that moved last week is
 * waste), never past `capMs`; back to the interval on any change, refusal or outage — those are what a puller is for.
 */
export function nextPullDelayMs(input: { outcome: PullBundleResult["status"]; unchangedStreak: number; intervalMs: number; capMs?: number }): number {
  const cap = input.capMs ?? 5 * 60 * 1000;
  if (input.outcome !== "unchanged") return input.intervalMs;
  const stretched = input.intervalMs * 2 ** Math.min(input.unchangedStreak, 20);
  return Math.min(Math.max(input.intervalMs, stretched), Math.max(cap, input.intervalMs));
}

export async function pullBundle(input: PullBundleInput): Promise<PullBundleResult> {
  const now = input.now();
  // What the caller handed back is returned unchanged on every failure: an ETag advanced past an answer the origin
  // never confirmed would make the next pull a 304 and hide the promotion until the one after it.
  const given: PullEdgeState = { ...NO_EDGE, ...(input.edge ?? {}) };
  const edge: PullEdgeState = { ...given };
  if (input.distributionPublicKey === null && input.scope.target !== "dev") return { status: "refused", reason: "plaintext_not_allowed", edge };
  const days = input.notAfterDays ?? 90;
  if (!Number.isFinite(days) || days <= 0 || days > 365) throw new Error("notAfterDays must be a positive number of days, 365 at most");
  const held = input.minimumGeneration ?? 0;

  let trustedRoot = input.trustedRoot;
  try {
    // The pointer first: unsigned and cacheable, it can only say "nothing moved" — never extend trust. A 304, or a
    // generation the caller already holds, ends the pull at the CDN. Anything else (moved, unknown, unreachable,
    // malformed) goes on to the origin, which is what a pointer can never replace — and so does a pointer that has
    // said "nothing moved" for longer than `maxPointerAgeMs`, the bound on a stuck one.
    const originAge = given.lastOriginAt ? Date.parse(now) - Date.parse(given.lastOriginAt) : Number.POSITIVE_INFINITY;
    const pointerTrusted = originAge <= (input.maxPointerAgeMs ?? DEFAULT_MAX_POINTER_AGE_MS);
    let pointerEtag: string | null | undefined;
    if (given.pointerUrl && !input.skipPointer && pointerTrusted) {
      try {
        const pointer = await input.client.edgePointer(given.pointerUrl, given.pointerEtag);
        if (pointer.status === "not_modified") return { status: "unchanged", via: "pointer", edge: given };
        if (pointer.status === "ok") {
          const generation = pointer.pointer?.generation;
          pointerEtag = pointer.etag;
          if (held > 0 && typeof generation === "number" && Number.isFinite(generation) && generation <= held) return { status: "unchanged", via: "pointer", edge: { ...given, pointerEtag: pointer.etag } };
        }
      } catch {
        // A CDN outage or a malformed pointer is not an answer: the origin is asked.
      }
    }
    if (input.fetchRoot) {
      const candidate = await input.fetchRoot();
      if (candidate) {
        const verdict = verifyRootMetadata({ candidate, trusted: trustedRoot, now });
        // A root that does not descend from the trusted one is the finding, not a detail behind unknown_signing_key.
        if (!verdict.ok) return { status: "refused", reason: "root_refused", detail: verdict.reason, edge: given };
        trustedRoot = candidate;
      }
    }
    // Conditional: the origin answers 304 to the ETag it last gave, and names the pointer either way.
    const fetched = await input.client.manifest({ ifNoneMatch: given.manifestEtag });
    if (fetched.status === "not_found") return { status: "nothing_promoted", edge: given };
    if (fetched.status === "unauthorized") return { status: "unavailable", reason: "unauthorized", edge: given };
    if (fetched.status === "forbidden") return { status: "unavailable", reason: "forbidden", ...(fetched.code ? { detail: fetched.code } : {}), edge: given };
    if (fetched.status === "error") return { status: "unavailable", reason: `http_${fetched.httpStatus}`, edge: given };
    // The origin answered: the pointer it names, the ETag it moved to, and the moment — the pointer's trust starts here.
    if (fetched.edgePointerUrl) edge.pointerUrl = fetched.edgePointerUrl;
    if (pointerEtag !== undefined) edge.pointerEtag = pointerEtag;
    edge.lastOriginAt = now;
    if (fetched.status === "not_modified") return { status: "unchanged", via: "origin", edge };
    const manifest = fetched.manifest;
    if (manifest.payload.generation < held) return { status: "refused", reason: "generation_rollback", held, detail: `the control plane answered generation ${manifest.payload.generation}; the caller holds ${held}`, edge: given };

    const payloads = new Map<string, Uint8Array>();
    for (const [hash, byteLength] of referencedPayloads(manifest.payload)) {
      const bytes = await input.client.payload(hash);
      if (!bytes) return { status: "refused", reason: "payload_missing", contentHash: hash, edge: given };
      if (bytes.length !== byteLength) return { status: "refused", reason: "payload_length_mismatch", contentHash: hash, edge: given };
      payloads.set(hash, bytes);
    }
    // No stored generation here: a puller has no host to move backwards. Anti-rollback is the store's rule, applied
    // by every runtime that opens this bundle.
    const verdict = verifyManifest({ manifest, root: trustedRoot, now, scope: input.scope, storedGeneration: 0, payloads, countersignRoot: input.countersignRoot ?? null, ...(input.requireCountersign !== undefined ? { requireCountersign: input.requireCountersign } : {}) });
    if (!verdict.ok) return { status: "refused", reason: verdict.reason, edge: given };
    // The bundle is built: the manifest's ETag is the caller's to keep — with the row, never before it.
    edge.manifestEtag = fetched.etag;

    const notAfter = new Date(Date.parse(now) + days * 86_400_000).toISOString();
    const contents: BundleContents = {
      createdAt: now,
      notAfter,
      manifest,
      keySet: trustedRoot,
      payloads: [...payloads].map(([contentHash, bytes]) => ({ contentHash: contentHash as `sha256:${string}`, byteLength: bytes.length, bytes: Buffer.from(bytes).toString("base64url") })),
    };
    const bundle = input.distributionPublicKey ? createEncryptedBundle(contents, input.distributionPublicKey) : createPlaintextBundle(contents);
    return { status: "ok", bundle, manifest, generation: manifest.payload.generation, releaseDigest: manifest.payload.releaseDigest, createdAt: now, notAfter, trustedRoot, edge };
  } catch (error) {
    return { status: "unavailable", reason: "network", detail: (error as Error).message, edge: given };
  }
}
